File: camera/ghostcam/commands.py
"""Server-issued command dispatch.

Mirrors camera/commands.go originally. Many commands cause the process
to exit so systemd restarts it with new state — same approach as the
Go camera. The new power-mode commands (set_power_mode,
set_upload_mode, set_schedule, set_battery_rules, upload_segments)
take effect WITHOUT a restart by mutating PowerModeState and the
upload queue in-process.
"""

from __future__ import annotations

import asyncio
import logging
import os
from collections.abc import Callable
from pathlib import Path

from ghostcam.client import Client
from ghostcam.config import write_stored_file
from ghostcam.credentials import clear_credentials
from ghostcam.firmware import check_firmware_update
from ghostcam.platform import ensure_wifi
from ghostcam.power_mode import PowerModeState
from ghostcam.wire import CameraCommand

logger = logging.getLogger(__name__)


# `prioritize_uploads` is the upload loop's hook for `upload_segments`
# commands — given a list of segment IDs, push them to the front of
# the upload queue ahead of normal new-segment work.
PriorityUploadCallback = Callable[[list[str]], None]
# Holder setter / getter for the current recording_mode. When wired up,
# set_recording_mode commands can hot-swap constant ↔ motion without a
# daemon restart.
RecordingModeSetter = Callable[[str], None]
RecordingModeGetter = Callable[[], str]


async def handle_command(
    cmd: CameraCommand,
    data_dir: Path,
    client: Client,
    *,
    power: PowerModeState | None = None,
    prioritize_uploads: PriorityUploadCallback | None = None,
    set_recording_mode: RecordingModeSetter | None = None,
    get_recording_mode: RecordingModeGetter | None = None,
) -> None:
    """Dispatch a single CameraCommand.

    `power` and `prioritize_uploads` are optional so existing call sites
    in tests / partial setups still work, but the daemon's main wires
    them in for the new commands to actually take effect.
    """
    if cmd.type == "reboot":
        logger.info("reboot command received")
        os._exit(0)
    elif cmd.type == "unregister":
        logger.info("unregister command received, clearing credentials")
        clear_credentials(data_dir)
        os._exit(0)
    elif cmd.type == "set_recording_mode":
        if cmd.mode is None:
            return
        new_mode = cmd.mode
        logger.info("recording mode change requested: %s", new_mode)
        try:
            write_stored_file(data_dir, "recording_mode", new_mode)
        except OSError as e:
            logger.error("failed to persist recording_mode: %s", e)
            return
        # Hot-swap when only the upload-decision flag changes (constant
        # ↔ motion). The upload loop reads via get_recording_mode each
        # iteration. Transitions to/from "never" still require a restart
        # because the watcher and upload tasks are conditionally spawned
        # in main.py based on the boot-time mode.
        current_mode = get_recording_mode() if get_recording_mode is not None else None
        if (
            set_recording_mode is not None
            and current_mode is not None
            and current_mode != "never"
            and new_mode != "never"
        ):
            set_recording_mode(new_mode)
            logger.info("recording mode hot-swapped: %s → %s", current_mode, new_mode)
            return
        logger.info("recording mode updated, restarting to apply")
        os._exit(0)
    elif cmd.type == "set_resolution":
        if cmd.resolution is None:
            return
        logger.info("resolution change requested: %s", cmd.resolution)
        try:
            write_stored_file(data_dir, "resolution", cmd.resolution)
        except OSError as e:
            logger.error("failed to persist resolution: %s", e)
            return
        logger.info("resolution updated, restarting to apply")
        os._exit(0)
    elif cmd.type == "network_config":
        logger.info("network config command: ssid=%s", cmd.ssid)
        if cmd.ssid:
            asyncio.create_task(_safe_ensure_wifi(cmd.ssid, cmd.psk))
    elif cmd.type == "update_firmware":
        logger.info("firmware update command received")
        if await check_firmware_update(client, data_dir):
            os._exit(0)
    elif cmd.type == "remove_network":
        logger.info("remove network command: ssid=%s", cmd.ssid)

    # --- power-mode commands ---
    elif cmd.type == "set_power_mode":
        if not cmd.power_mode or power is None:
            return
        logger.info("power_mode change requested: %s", cmd.power_mode)
        try:
            power.set_manual_power_mode(cmd.power_mode)
        except ValueError as e:
            logger.warning("ignoring invalid power_mode: %s", e)
    elif cmd.type == "set_upload_mode":
        if not cmd.upload_mode or power is None:
            return
        logger.info("upload_mode change requested: %s", cmd.upload_mode)
        try:
            power.set_manual_upload_mode(cmd.upload_mode)
        except ValueError as e:
            logger.warning("ignoring invalid upload_mode: %s", e)
    elif cmd.type == "set_schedule":
        if power is None:
            return
        # Empty string is "clear schedule".
        body = cmd.schedule or "[]"
        logger.info("schedule update received (%d chars)", len(body))
        try:
            power.set_schedule(body)
        except (ValueError, KeyError, TypeError) as e:
            logger.warning("ignoring malformed schedule: %s", e)
    elif cmd.type == "set_battery_rules":
        if power is None:
            return
        body = cmd.battery_rules or "[]"
        logger.info("battery rules update received (%d chars)", len(body))
        try:
            power.set_battery_rules(body)
        except (ValueError, KeyError, TypeError) as e:
            logger.warning("ignoring malformed battery rules: %s", e)
    elif cmd.type == "upload_segments":
        if not cmd.segment_ids:
            return
        logger.info(
            "upload_segments command: %d segment(s) requested",
            len(cmd.segment_ids),
        )
        if prioritize_uploads is None:
            logger.warning(
                "upload_segments received but no upload-prioritise hook wired in"
            )
            return
        prioritize_uploads(list(cmd.segment_ids))

    else:
        logger.warning("unknown command type: %s", cmd.type)


async def _safe_ensure_wifi(ssid: str, psk: str | None) -> None:
    try:
        await ensure_wifi(ssid, psk)
    except Exception as e:  # noqa: BLE001
        logger.warning("WiFi config failed: %s", e)

File: camera/internal/commands/commands.go
package commands

import (
	"context"
	"encoding/json"
	"log/slog"
	"os"
	"time"

	"github.com/cargocam/ghostcam/camera/internal/battery"
	"github.com/cargocam/ghostcam/camera/internal/diag"
	"github.com/cargocam/ghostcam/camera/internal/firmware"
	"github.com/cargocam/ghostcam/camera/internal/network"
	"github.com/cargocam/ghostcam/camera/internal/power"
	"github.com/cargocam/ghostcam/camera/internal/state"
	"github.com/cargocam/ghostcam/camera/internal/uplink"
	"github.com/cargocam/ghostcam/common"
)

// Client is the command dispatcher's narrow view of the server HTTP
// client. Currently only the firmware updater needs it; the surface is
// firmware.Client + a Version method so main can pass Version through
// without each subpackage reaching back into package main. The
// concrete *main.Client (camera/client.go) satisfies this surface.
type Client interface {
	firmware.Client
	Version() string
}

// HandleCommand processes a server-issued command received via the telemetry
// poll response. Commands like reboot, unregister, and config changes cause
// the process to exit — systemd restarts it with new state.
func HandleCommand(ctx context.Context, cmd common.CameraCommand, dataDir string, client Client) {
	switch cmd.Type {
	case "reboot":
		slog.Info("reboot command received")
		os.Exit(0)
	case "unregister":
		slog.Info("unregister command received, clearing credentials")
		state.ClearCredentials(dataDir)
		os.Exit(0)
	case "set_recording_mode":
		slog.Info("recording mode change requested", "mode", cmd.Mode)
		if err := state.WriteStoredFile(dataDir, "recording_mode", cmd.Mode); err != nil {
			slog.Error("failed to persist recording_mode", "err", err)
			return
		}
		slog.Info("recording mode updated, restarting to apply")
		os.Exit(0)
	case "set_resolution":
		slog.Info("resolution change requested", "resolution", cmd.Resolution)
		if err := state.WriteStoredFile(dataDir, "resolution", cmd.Resolution); err != nil {
			slog.Error("failed to persist resolution", "err", err)
			return
		}
		slog.Info("resolution updated, restarting to apply")
		os.Exit(0)
	case "set_power_mode":
		// Power-mode changes are handled in-process (no os.Exit) because
		// the sleep mode needs telemetry to keep flowing so the wake
		// command can arrive — restarting the daemon every time would
		// break that loop.
		slog.Info("power mode change requested", "mode", cmd.PowerMode)
		if !power.IsValidPowerMode(cmd.PowerMode) {
			slog.Error("invalid power mode, ignoring", "mode", cmd.PowerMode)
			return
		}
		if err := state.WriteStoredFile(dataDir, "power_mode", cmd.PowerMode); err != nil {
			slog.Error("failed to persist power_mode", "err", err)
			return
		}
		power.SetManualPowerMode(cmd.PowerMode)
		previous, effective := power.ApplyEffectivePowerMode(battery.LastBatteryPct())
		if previous != effective {
			// Tear down any in-flight capture session so the loop's
			// top-of-iteration check can re-evaluate the mode. Same
			// mechanism the telemetry-poll uses on WHIPSessionMissing
			// and ABR uses on tier shift.
			state.RequestPipelineRestart()
			slog.Info("power mode updated",
				"from", previous, "to", effective, "manual", cmd.PowerMode)
		}
	case "set_battery_rules":
		// Battery rules are level-triggered against the most recent
		// battery_pct sample. Persisting them and swapping the in-process
		// atomic is enough — no pipeline restart, because the rules only
		// take effect on the next telemetry tick, where the standard
		// ApplyEffectivePowerMode path will fire requestPipelineRestart
		// if the effective mode changed.
		var rules []battery.BatteryRule
		if cmd.BatteryRules != "" {
			if err := json.Unmarshal([]byte(cmd.BatteryRules), &rules); err != nil {
				slog.Error("set_battery_rules: invalid JSON", "err", err, "payload", cmd.BatteryRules)
				return
			}
		}
		if err := battery.SaveBatteryRules(dataDir, rules); err != nil {
			slog.Error("set_battery_rules: persist failed", "err", err)
			return
		}
		battery.SetBatteryRules(rules)
		previous, effective := power.ApplyEffectivePowerMode(battery.LastBatteryPct())
		if previous != effective {
			state.RequestPipelineRestart()
			slog.Info("battery rules updated, effective mode changed",
				"from", previous, "to", effective, "rule_count", len(rules))
		} else {
			slog.Info("battery rules updated", "count", len(rules))
		}
	case "network_config":
		slog.Info("network config command", "ssid", cmd.SSID)
		go func() {
			psk := cmd.PSK
			if err := network.EnsureWifi(ctx, cmd.SSID, &psk); err != nil {
				slog.Warn("WiFi config failed", "err", err)
			}
		}()
	case "set_cellular":
		// Deliver/refresh the cellular APN on a deployed camera without
		// SSH. Persist so it survives reboots (LoadConfig reads the file),
		// then apply immediately in the background (nmcli can block on a
		// cold modem). Empty APN is a no-op guard.
		slog.Info("set cellular command", "apn", cmd.CellularAPN)
		if cmd.CellularAPN == "" {
			break
		}
		state.PersistCellular(dataDir, cmd.CellularAPN, cmd.CellularUser, cmd.CellularPass)
		go func() {
			if err := network.EnsureCellular(ctx, cmd.CellularAPN, cmd.CellularUser, cmd.CellularPass); err != nil {
				slog.Warn("set_cellular apply failed", "err", err)
			}
		}()
	case "force_cellular":
		// Dev/diagnostic: force the camera onto cellular by taking WiFi
		// down for N seconds, then auto-restore. We only persist the
		// deadline here — the uplink watchdog (started in main) does the
		// enforcement and the reboot-surviving revert, so a failed cellular
		// link can't strand the camera. seconds<=0 reverts immediately.
		slog.Info("force cellular command", "seconds", cmd.ForceCellularSeconds)
		uplink.SetForce(dataDir, time.Now().UnixMilli(), cmd.ForceCellularSeconds)
	case "update_firmware":
		slog.Info("firmware update command received")
		if firmware.CheckFirmwareUpdate(ctx, client, dataDir, client.Version()) {
			os.Exit(0)
		}
	case "remove_network":
		slog.Info("remove network command", "ssid", cmd.SSID)
	case "diag_bundle":
		// Capture happens in a goroutine so HandleCommand returns
		// promptly and the telemetry-poll dispatcher can keep moving
		// through any other queued commands. The bundle lands in
		// pendingDiagBundles and rides out on the next poll.
		slog.Info("diag_bundle command received", "diag_id", cmd.DiagID)
		diagID := cmd.DiagID
		go func() {
			bundle := diag.CaptureDiagBundle(context.Background(), diagID)
			diag.AddPendingDiagBundle(bundle)
			slog.Info("diag_bundle captured, queued for next poll", "diag_id", diagID)
		}()
	case "restart_service":
		// os.Exit; systemd respawns us. No goroutine — we want to
		// take effect immediately.
		RestartGhostcamService()
	case "restart_modem_manager":
		go RestartModemManager(context.Background())
	case "restart_network_manager":
		go RestartNetworkManager(context.Background())
	default:
		slog.Warn("unknown command", "type", cmd.Type)
	}
}
